zer0-os/zOS
# fix(app-sandbox): scrollbar width

### What does this do?

Adds functionality to prevent the sandbox scrollbar from pushing sandbox content to the left.

https://user-images.githubusercontent.com/12437916/198920015-09c35b8a-1f88-4a4a-ac0f-3ae70a577de7.mov

**Note: this is not an issue on Firefox.**

- Added a `ref` to the `app-sandbox` element.
- Add a `ResizeObserver` to `app-sandbox` to handle size changes. I attached a `ResizeObserver` to `app-sandbox` instead of using a `window` event listener because the context menus opening/closing could cause the app to overflow (and add a scrollbar).
- On `app-sandbox` size change, calculate scroll bar width and apply scroll bar width as `paddingLeft` on `app-sandbox`. Scroll bar width should be calculated as browsers seem to all react differently to scrollbar styling, and because it allows for more robust code at the cost of a tiny computation (i.e. don't have to hardcode the scrollbar width).

### Why are we making this change?

Because the app shouldn't be pushed left when the `app-sandbox` overflows.

### How do I test this?

1. Open the current development build in any non-Firefox browser.
2. Open any app which overflows (I used `Trade`).
3. Zoom out until the content no longer overflows.
4. Resize your browser (vertical) until the scroll bar appears (and pushes the content left).


### Key decisions and Risk Assessment:
  #### Things to consider:
  1. How will this affect security?
-

  2. How will this affect performance?
`ResizeObserver` is about as fast as we're going to get this. The `app-sandbox` container only sets state when scrollbar width changes, so shouldn't happen in quick succession.

  3. Does this change any APIs?
Added an extra prop to `app-sandbox`, and an extra value to the `layout` context. Apart from that - no changes.
